### PR TITLE
[spirv] Add support for four register number shifting options

### PR DESCRIPTION
**Pull request by antiagainst**
Added the following four command line options to shift register
number for Vulkan:

* -fvk-b-shift
* -fvk-t-shift
* -fvk-s-shift
* -fvk-u-shift

These options are used to avoid assigning the same binding
number to more than one resources.

### CODE REVIEW

**Comment by antiagainst** [comment]
@marcelolr: This pull request added 4 command line options, so it would benefit from your review. To make it easier, all modifications outside the normal SPIR-V directories are gathered in the first commit. :)

**Comment by antiagainst** [comment]
FYI @chaoticbob

File: lib/DxcSupport/HLSLOptions.cpp [hunk 1]
```diff
@@ -444,15 +444,42 @@ int ReadDxcOpts(const OptTable *optionTable, unsigned flagsToInclude,
...
+    errors << "invalid s-type register number shift: " << sShift;
+    return 1;
+  }
+  const llvm::StringRef uShift = Args.getLastArgValue(OPT_fvk_u_shift_EQ, "0");
+  if (uShift.getAsInteger(10, opts.VkUShift)) {
+    errors << "invalid u-type register number shift: " << uShift;
+    return 1;
+  }
+
   opts.VkStageIoOrder = Args.getLastArgValue(OPT_fvk_stage_io_order_EQ, "decl");
```
Comment (marcelolr) [inline lib/DxcSupport/HLSLOptions.cpp]: You might also want to provide a message to tell users that the register shifts have no effect unless SPIR-V is being generated if they did in fact specify any of the shift options but missed the overall SPIR-V flag.

File: include/dxc/Support/HLSLOptions.td [hunk 2]
```diff
@@ -234,10 +234,18 @@ def no_min_precision: Flag<["-", "/"], "no-min-precision">, Flags<[CoreOption, D
 def ignore_line_directives : Flag<["-", "/"], "ignore-line-directives">, HelpText<"Ignore line directives">, Flags<[CoreOption]>, Group<hlslcomp_Group>;
 
 // SPIRV Change Starts
-def spirv : Flag<["-"], "spirv">, Group<spirv_Group>, Flags<[CoreOption, DriverOption, HelpHidden]>,
+def spirv : Flag<["-"], "spirv">, Group<spirv_Group>, Flags<[CoreOption, DriverOption]>,
   HelpText<"Generate SPIR-V binary code">;
 def fvk_stage_io_order_EQ : Joined<["-"], "fvk-stage-io-order=">, Group<spirv_Group>, Flags<[CoreOption, DriverOption, HelpHidden]>,
   HelpText<"Specify Vulkan stage I/O location assignment order">;
+def fvk_b_shift_EQ : Joined<["-"], "fvk-b-shift=">, Group<spirv_Group>, Flags<[CoreOption, DriverOption]>,
```
Comment (marcelolr) [inline include/dxc/Support/HLSLOptions.td]: I'd re-emphasize the SPIR-V-ness of these options with something along the lines of "...register number shift for SPIR-V b-type register" (same for other flags).

Comment (antiagainst) [reply include/dxc/Support/HLSLOptions.td]: Done! Changed to `Specify Vulkan binding number shift for t-type register` and similar.

File: lib/DxcSupport/HLSLOptions.cpp (hunk 1 above)
Comment (antiagainst) [reply lib/DxcSupport/HLSLOptions.cpp]: Good point! I've made it an error if `-spirv` is missing but `-fvk-*-shift` is provided.

**Pull request merged**